tempesta_fw/t/unit/test_tfw_str.c

```diff
@@ -101,102 +102,130 @@ TEST(tfw_strcpy, dst_compound)
 
 TEST(tfw_strcpy, both_compound)
 {
-	char buf[32] = { [0 ... 30] = 'a', 0 };
-	TFW_STR(s1, buf);
+	DEFINE_TFW_STR(s1, "abcdefghijklmnop");
+
 	TfwStr s2 = {
-		.ptr = (TfwStr []){
-			{ .ptr = "ab",	.len = 2 },
-			{ .ptr = "cde",	.len = 3 },
-			{ .ptr = "f",	.len = 1 },
-			{ .ptr = "g",	.len = 1 },
-			{ .ptr = "h",	.len = 1 },
-			{ .ptr = "ijklmno", .len = 7 },
-			{ .ptr = "p", .len = 1 }
+		.chunks = (struct TfwStr *)(TfwStr []){
+			{ .data = "ab",	.len = 2 },
+			{ .data = "cde",	.len = 3 },
+			{ .data = "f",	.len = 1 },
+			{ .data = "g",	.len = 1 },
+			{ .data = "h",	.len = 1 },
+			{ .data = "ijklmno", .len = 7 },
+			{ .data = "p", .len = 1 }
 		},
 		.len = sizeof("abcdefghijklmnop") - 1,
-		.flags = 7 << TFW_STR_CN_SHIFT
+		.flags = 0
 	};
-
-	EXPECT_ZERO(tfw_strcpy(s1, &s2));
-	EXPECT_TRUE(tfw_str_eq_cstr(s1, "abcdefghijklmnop",
+	TFW_DBG("test:both:cpy\n");
+	EXPECT_ZERO(tfw_strcpy(&s1, &s2));
+	EXPECT_TRUE(tfw_str_eq_cstr(&s1, "abcdefghijklmnop",
 				    sizeof("abcdefghijklmnop") - 1, 0));
 }
 
 TEST(tfw_strcat, plain)
 {
-	int chunks;
-	TFW_STR(s1, "abcdefghijklmnop");
+	int chunkscnt;
+/*	TfwStr s1 = {
+		.chunks = (struct TfwStr *)(TfwStr []){
+			{ .data = "ab",	.len = 2 },
+			{ .data = "cde",	.len = 3 },
+			{ .data = "f",	.len = 1 },
+			{ .data = "g",	.len = 1 },
+			{ .data = "h",	.len = 1 },
+			{ .data = "ijklmno", .len = 7 },
+			{ .data = "p", .len = 1 }
+		},
+		.len = sizeof("abcdefghijklmnop") - 1,
+		.flags = 0
+	};*/
+	TfwStr s1 = {.data = "abcdefghijklmnop", .len = sizeof("abcdefghijklmnop") - 1,.skb = NULL};
+/*
+	TfwStr s1 = { .chunks = (struct TfwStr *)(TfwStr []){{
+			.data = "abcdefghijklmnop", .len = sizeof("abcdefghijklmnop") - 1,.skb = NULL}},
+		.len = sizeof("abcdefghijklmnop") - 1,
+		.skb = NULL,
+		.flags = 0
+		   };
+*/
 	TfwStr s2 = {
 		.len = sizeof("0123456789") - 1,
-		.ptr = "0123456789"
+		.data = "0123456789"
```

Comment:
There are macros `DEFINE_TFW_STR()` and `TFW_STR_FROM()` that can be used for this purpose.